### PR TITLE
feat: extract per-episode topic tags during summarization

### LINKED ISSUE #259: feat: Extract per-episode topic tags during summarization
Part of #258

## Context

Currently, topic information only exists as ephemeral daily snapshots in the `trending_topics` table — a JSON blob generated by `generate-trending-topics.ts` that clusters recent episodes by theme. This data isn't queryable per-episode and can't be used for cross-referencing or scoring.

This sub-issue adds structured, per-episode topic tags extracted during summarization and stored in a normalized junction table. This is the foundation that enables cross-episode ranking (sub-issue #3) and personal relevance scoring (sub-issue #4).

## Complexity Hint

**Subsystems touched:** DB schema, summarization prompt, Trigger.dev helpers, SummaryResult type
**Independent stories:** 1
**Structural changes needed:** Yes — new `episode_topics` table, `SummaryResult` type extension, prompt output schema change
**Scope note:** None detected

## Task

Add topic tag extraction to the summarization pipeline and persist tags in a new `episode_topics` table.

**Topic tag format:** Each episode gets 1-5 topic tags, each with a canonical name (2-5 words, e.g. "AI Ethics", "Leadership") and a relevance score (0.0-1.0, how central the topic is to the episode).

**Prompt changes:** Extend the summarization prompt to request a `topics` array in the JSON output. The LLM should extract topics from the transcript content itself, using concise professional labels.

**Storage:** New normalized `episode_topics` junction table — not a JSON column on `episodes` — so topics are queryable across episodes.

## Acceptance Criteria

- [ ] VERIFY: New `episode_topics` table exists with columns: `id`, `episodeId` (FK → episodes), `topic` (text), `relevance` (decimal 0.00-1.00), `createdAt`
- [ ] VERIFY: `episode_topics` has an index on `topic` for efficient cross-episode queries
- [ ] VERIFY: `episode_topics` has a unique index on `(episodeId, topic)` to prevent duplicates
- [ ] VERIFY: `SummaryResult` type includes optional `topics: Array<{ name: string; relevance: number }>` field
- [ ] VERIFY: Summarization prompt requests topic extraction in its JSON output format
- [ ] VERIFY: `generateEpisodeSummary()` validates topic output (array of objects with string name and numeric relevance)
- [ ] VERIFY: `persistEpisodeSummary()` writes topic tags to `episode_topics` table
- [ ] VERIFY: Missing or malformed topics from LLM don't break summarization (graceful degradation — store summary without topics)
- [ ] VERIFY: Existing summarized episodes without topics continue to work (no migration required for old data)
- [ ] VERIFY: Unit tests cover: topic extraction parsing, validation, persistence, graceful fallback
- [ ] VERIFY: `bun run lint`, `bun run test`, and `bun run build` pass

## Edge Cases

- LLM returns no topics (empty array) — store summary normally, skip `episode_topics` insert
- LLM returns duplicate topic names for same episode — deduplicate before insert
- LLM returns topics with relevance outside 0-1 range — clamp to 0.0-1.0
- LLM returns more than 5 topics — take top 5 by relevance
- Topic names with inconsistent casing ("AI Ethics" vs "ai ethics") — normalize to title case before storage

### Files to Modify

- `src/db/schema.ts` — add `episodeTopics` table definition, relations, type exports
- `src/lib/prompts.ts` — extend `getSummarizationPrompt()` JSON format and instructions to include `topics` array
- `src/lib/openrouter.ts` — add optional `topics` field to `SummaryResult` interface
- `src/trigger/helpers/ai-summary.ts` — validate and normalize topic extraction output from LLM
- `src/trigger/helpers/database.ts` — extend `persistEpisodeSummary()` to write `episode_topics` rows
- `src/trigger/helpers/__tests__/ai-summary.test.ts` — test topic validation/normalization
- `src/trigger/helpers/__tests__/database.test.ts` — test topic persistence

### Approach

1. Add `episodeTopics` table to `src/db/schema.ts` with appropriate columns, indexes, relations, and type exports. Run `bun run db:generate` to create migration.
2. Extend `SummaryResult` in `src/lib/openrouter.ts` with optional `topics` field.
3. Update `getSummarizationPrompt()` in `src/lib/prompts.ts` to request topic extraction in the JSON output. Add instructions for concise professional topic labels and relevance scoring.
4. Update `generateEpisodeSummary()` in `src/trigger/helpers/ai-summary.ts` to validate and normalize topic output: enforce array shape, clamp relevance to 0-1, deduplicate, limit to 5, normalize casing.
5. Update `persistEpisodeSummary()` in `src/trigger/helpers/database.ts` to insert rows into `episode_topics` after the episode update. Use a transaction. Skip insert if topics are empty/undefined.
6. Add unit tests for topic validation, normalization, and persistence.
7. Run `bun run db:push` to apply schema to dev database.
8. Test end-to-end by summarizing an episode and verifying topic tags appear in the database.

### Patterns to Follow

- Follow the existing `trendingTopics` table pattern in `src/db/schema.ts` for table definition style.
- Follow the existing validation pattern in `ai-summary.ts` lines 40-54 (the `worthItDimensions` validation) for topic validation.
- Follow the existing `persistEpisodeSummary()` pattern in `database.ts` for database writes.

### Constraints

- Topic extraction must not break existing summarization — if the LLM doesn't return topics, the rest of the summary must still be processed and stored.
- No migration needed for existing episodes — topics are optional and only populated going forward.
- Custom prompts (via `aiConfig.summarizationPrompt`) should still work without topics.

### In Scope
- New `episode_topics` table and schema
- Extending summarization prompt to extract topics
- Validating and persisting topic tags
- Unit tests

### Out of Scope
- Backfilling topics for already-summarized episodes (separate task if needed)
- UI display of topic tags (comes with later sub-issues)
- Replacing the existing trending topics system (can be done later as the new data subsumes it)
- Cross-episode comparison using topics (sub-issue #3)

## Dependencies

- Blocked by: None
- Blocks: #258 sub-issues #3 and #4

## Definition of Done

- [ ] Implementation complete
- [ ] Tests passing (unit + integration)
- [ ] `bun run lint`, `bun run test`, `bun run build` pass
- [ ] PR approved and merged
- [ ] Epic checklist updated

### PR DESCRIPTION
**Pull request by funny-otter**
## Summary

- Add topic tag extraction to the AI summarization pipeline — the LLM now returns 1-5 topic tags per episode with relevance scores
- New `episode_topics` junction table with indexed `topic` column for cross-episode queries and unique `(episodeId, topic)` constraint
- Topic validation pipeline: filter invalid entries, title-case normalization, relevance clamping (0-1), deduplication (highest relevance wins), cap at 5
- Persistence in both update and insert paths of `persistEpisodeSummary()` with `onConflictDoNothing` for idempotent retries
- ADR-031 documents the junction table vs JSON column decision

## Architecture Decisions

- **Junction table over JSON column** — enables indexed cross-episode queries (`WHERE topic = ?`) without JSON parsing
- **No transaction** — benign failure mode (summary saved without topics), Trigger.dev retries self-heal
- **Custom prompts bypass** — `interpolatePrompt` path doesn't inject topic extraction; `topics` field is optional on `SummaryResult`
- **`onConflictDoNothing`** — Trigger.dev task retries won't fail or duplicate

## Deploy Notes

> **IMPORTANT:** Run `doppler run --config prd -- bunx drizzle-kit push` against production **before** this code ships. The `episode_topics` table must exist before `persistEpisodeSummary` runs. See ADR-031 and the `worth_it_reason` column incident for prior art.

## Files Changed

| File | Change |
|------|--------|
| `src/db/schema.ts` | `episodeTopics` table, relations, type exports |
| `src/lib/openrouter.ts` | `SummaryResult.topics` optional field |
| `src/lib/prompts.ts` | Topic extraction in summarization prompt |
| `src/trigger/helpers/ai-summary.ts` | `toTitleCase` helper, validation/normalization pipeline |
| `src/trigger/helpers/database.ts` | Topic persistence (update + insert paths) |
| `src/trigger/helpers/__tests__/ai-summary.test.ts` | 13 topic normalization tests |
| `src/trigger/helpers/__tests__/database.test.ts` | 5 topic persistence tests + existing test updates |
| `docs/adr/031-episode-topics-junction-table.md` | ADR for schema decision |
| `drizzle/0016_abnormal_wendell_vaughn.sql` | Migration (additive only) |

## Known Limitations

- `toTitleCase` collapses acronyms: "AI Ethics" → "Ai Ethics". Acceptable for v1; follow-up issue for acronym-preserving normalization.
- Custom prompts don't automatically get topic extraction — this is intentional (AD-3 in ADR-031).

## Test Plan

- [x] `bun run lint` — clean
- [x] `bun run test` — 1507 tests passing (129 files)
- [x] `bun run build` — clean
- [x] Migration SQL reviewed — additive only (CREATE TABLE, indexes, FK), no DROPs
- [ ] Run `bun run db:push` against production before deploy
- [ ] Summarize an episode end-to-end and verify topic tags appear in `episode_topics` table

Closes #259

## Summary by CodeRabbit

* **New Features**
  * Episode summaries now extract up to 5 normalized, deduplicated topic tags (Title Case) with relevance scores (0.00–1.00) and persist them per episode.

* **Behavioral Change**
  * Summarization prompt now requests a top-level topics field; custom prompts bypass automatic topic extraction. Topics may be reconciled on retries.

* **Documentation**
  * Added ADR detailing topic persistence and reconciliation approach.

* **Tests**
  * Added tests for topic normalization and persistence scenarios.

### CODE REVIEW

File: src/trigger/helpers/ai-summary.ts [hunk 1]
```diff
@@ -53,6 +58,39 @@ export async function generateEpisodeSummary(
...
+    if (Array.isArray(result.topics)) {
+      result.topics = result.topics
+        // 1. Array shape check: drop entries missing name or relevance
+        .filter(
+          (t): t is { name: string; relevance: number } =>
+            typeof t === "object" &&
+            t !== null &&
+            typeof t.name === "string" &&
+            t.name.trim().length > 0 &&
+            typeof t.relevance === "number"
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/ai-summary.ts]: Fixed: added `!Number.isNaN(t.relevance)` guard in commit 60b7ca3.

File: src/trigger/helpers/database.ts [hunk 2]
```diff
@@ -134,6 +134,23 @@ export async function persistTranscript(
...
+  await db
+    .insert(episodeTopics)
+    .values(
+      topics.map((t) => ({
+        episodeId,
+        topic: t.name,
+        relevance: t.relevance.toFixed(2),
+      }))
+    )
+    .onConflictDoNothing(); // unique (episodeId, topic) — idempotent on re-runs
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/database.ts]: Fixed: replaced `onConflictDoNothing` with delete-then-insert to reconcile stale topics on re-summarization. Tests updated to verify the new behavior.

**Comment by funny-otter** [comment]
> 🧹 Nitpick comments (1) — src/db/schema.ts: Add DB-level bounds for `topic` to prevent blank/oversized entries...

Acknowledged — will be addressed by the author. App-level validation already filters blanks and the `toTitleCase` normalization keeps topic strings short, but adding `varchar` bounds and a `topic_not_blank` constraint at the DB level is reasonable defense-in-depth for a follow-up.

**Comment by funny-otter** [comment]
## PR Comment Status

| Status | Threads | Review Bodies | Issue Comments | Total |
|--------|---------|---------------|----------------|-------|
| Resolved | 1 | 2 | 9 | 12 |
| Fixed by DLC | 1 | 0 | 0 | 1 |
| Discussion-Deferred | 0 | 1 | 0 | 1 |
| **Total** | **2** | **3** | **9** | **14** |

## Decisions

- Inline thread: `src/trigger/helpers/database.ts` — Fixed: delete-then-insert to reconcile stale topics on re-summarization
- Review body: `pr_comment:4094150075` — Deferred: schema hardening nitpick (varchar bounds + topic_not_blank constraint) — app-level validation covers this; good candidate for follow-up

## Follow-up

Author will address the schema hardening nitpick in a future PR.
